### PR TITLE
Push notifications.

### PR DESCRIPTION
**Pull request by dbrant**

### CODE REVIEW

File: app/build.gradle [hunk 1]
```diff
@@ -201,6 +202,9 @@ dependencies {
     implementation 'com.github.chrisbanes:PhotoView:2.3.0'
     implementation "com.github.skydoves:balloon:1.2.1"
 
+    // For receiving push notifications for logged-in users.
+    implementation 'com.google.firebase:firebase-messaging:20.2.4'
```
Comment (sharvaniharan) [inline app/build.gradle:206]: Looks like 20.3.0 is available.

**Comment by dbrant** [comment]
@cooltey @sharvaniharan  This is now ready for review / merge.

File: app/src/extra/java/org/wikipedia/push/WikipediaFirebaseMessagingService.kt [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+                return
+            }
+
+            val token = Prefs.getPushNotificationToken()
+            val oldToken = Prefs.getPushNotificationTokenOld()
+
+            // Make sure to unsubscribe the previous token, if any
+            if (oldToken.isNotEmpty()) {
+                if (oldToken != token) {
+                    ServiceFactory.get(WikipediaApp.getInstance().wikiSite).unsubscribePush(csrfToken, Prefs.getPushNotificationToken())
```
Comment (cooltey) [inline app/src/extra/java/org/wikipedia/push/WikipediaFirebaseMessagingService.kt:106]: A little bit confused here. Should we unsubscribe the `token` or `oldToken`?

File: app/src/extra/java/org/wikipedia/push/WikipediaFirebaseMessagingService.kt [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+                    .subscribeOn(Schedulers.io())
+                    .observeOn(AndroidSchedulers.mainThread())
+                    .subscribe({
+                        L.d("Notification options successfully.")
+                    }, {
+                        L.e(it)
+                    })
+        }
+
+        fun unsubscribePush(csrfToken: String): Observable<MwQueryResponse> {
```
Comment (cooltey) [inline app/src/extra/java/org/wikipedia/push/WikipediaFirebaseMessagingService.kt:163]: Would it be possible to reuse this method in `subscribeWithCsrf` to unsubscribe token?

**Review by cooltey — COMMENTED** [review]
Received notifications after ~20 minutes. Have a few comments/questions inline.

File: app/src/extra/java/org/wikipedia/push/WikipediaFirebaseMessagingService.kt (hunk 2 above)
Comment (dbrant) [reply app/src/extra/java/org/wikipedia/push/WikipediaFirebaseMessagingService.kt]: Whoops, fixed!

File: app/src/extra/java/org/wikipedia/push/WikipediaFirebaseMessagingService.kt (hunk 3 above)
Comment (dbrant) [reply app/src/extra/java/org/wikipedia/push/WikipediaFirebaseMessagingService.kt]: Ok!

**Comment by sharvaniharan** [comment]
Screenshot_20201016-103116
Got that in 5 mins

**Comment by sharvaniharan** [comment]
Lgtm... @cooltey please merge when you are done with review.

**Pull request merged**